### PR TITLE
SW-5737 Improve robustness of shapefile import

### PR DESCRIPTION
**Pull request by sgrimm**
We recently got a complex shapefile that failed to import for a couple reasons.

The big one was that it used 4-dimensional coordinates (the fourth being a
non-spatial "M" value that's used to hold application-defined values) which caused
some of the geometry calculations to bomb out.

It also had a lot of complex borders between subzones, which, thanks to
floating-point precision limitations, caused the calculated zone and site
boundaries to have small holes in them.

Make the importer more robust to this kind of shapefile:

- Strip the Z and M dimensions, if present
- Use fixed-precision coordinates to avoid floating-point errors
- Remove small gaps between adjacent subzones when calculating zone and site
  boundaries

### CODE REVIEW

**Comment by sgrimm** [comment]
* **#2222** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2222?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sgrimm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteImporter.kt [hunk 1]
```diff
@@ -30,6 +35,15 @@ class PlantingSiteImporter(
     val temporaryPlotCountProperties = setOf("temporary")
   }
 
+  /**
+   * When importing, use fixed-precision coordinates with 8 decimal digits rather than the default
+   * floating-point precision. This helps avoid introducing slight errors to calculated geometries
+   * such as when zone boundaries are derived from subzone boundaries.
+   *
+   * 8 decimal digits works out to a resolution of less than 1 centimeter.
```
Comment (nickgraz) [inline src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteImporter.kt:43]: 👍

File: src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteImporter.kt [hunk 2]
```diff
@@ -213,6 +227,32 @@ class PlantingSiteImporter(
...
-      geometries.reduce { a, b -> a.union(b) }.toMultiPolygon()
+  /**
+   * Merges a set of geometries into a MultiPolygon. If two geometries are adjacent but have a tiny
+   * gap, e.g., due to floating-point precision limitations, the gap is eliminated.
+   */
+  private fun mergeToMultiPolygon(geometries: Collection<Geometry>): MultiPolygon {
+    return geometries
+        .reduce { a, b ->
+          val tolerance = GeometrySnapper.computeOverlaySnapTolerance(a, b)
+          a.union(GeometrySnapper(b).snapTo(a, tolerance))
```
Comment (nickgraz) [inline src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteImporter.kt:238]: That library has everything!

**Review by nickgraz — APPROVED** [review]
Looks good to me

**Pull request merged**